File: sql/plsql/Go/plsql_parser_base.go
package parser

import (
    "github.com/antlr4-go/antlr/v4"
)

// PlSqlParserBase implementation.
type PlSqlParserBase struct {
    *antlr.BaseParser
    _isVersion12 bool
    _isVersion11 bool
    _isVersion10 bool
    _lastUnitWasPlsql bool
}

var StaticConfig PlSqlParserBase

func init() {
    StaticConfig = PlSqlParserBase {
        _isVersion12: true,
        _isVersion11: true,
	    _isVersion10: true,
    }
}

func (p *PlSqlParserBase) reset() {
    p._lastUnitWasPlsql = false
    p.BaseParser.Reset()
}

func (p *PlSqlParserBase) setLastUnitPlsql() { p._lastUnitWasPlsql = true }
func (p *PlSqlParserBase) setLastUnitSql()   { p._lastUnitWasPlsql = false }
func (p *PlSqlParserBase) isLastUnitSql() bool    { return !p._lastUnitWasPlsql }
func (p *PlSqlParserBase) isLastUnitPlsql() bool  { return p._lastUnitWasPlsql }

func (p *PlSqlParserBase) isSolidusSeparator() bool {
    stream := p.GetTokenStream().(*antlr.CommonTokenStream)
    solidus := stream.LT(1)
    if solidus == nil || solidus.GetTokenType() != PlSqlParserSOLIDUS {
        return false
    }

    solidusLine := solidus.GetLine()

    prev := stream.LT(-1)
    if prev != nil && prev.GetTokenType() != antlr.TokenEOF && prev.GetLine() == solidusLine {
        return false
    }

    next := stream.LT(2)
    if next != nil && next.GetTokenType() != antlr.TokenEOF && next.GetLine() == solidusLine {
        return false
    }

    return true
}

func (p *PlSqlParserBase) isVersion12() bool {
    return StaticConfig._isVersion12;
}

func (p *PlSqlParserBase) setVersion12(value bool) {
    StaticConfig._isVersion12 = value;
}

func (p *PlSqlParserBase) isVersion11() bool {
    return StaticConfig._isVersion11;
}

func (p *PlSqlParserBase) setVersion11(value bool) {
    StaticConfig._isVersion11 = value;
}

func (p *PlSqlParserBase) isVersion10() bool {
    return StaticConfig._isVersion10;
}

func (p *PlSqlParserBase) setVersion10(value bool) {
    StaticConfig._isVersion10 = value;
}

func (p *PlSqlParserBase) IsNotNumericFunction() bool {
    stream := p.GetTokenStream().(*antlr.CommonTokenStream)
    lt1 := stream.LT(1)
    lt2 := stream.LT(2)
    if (lt1.GetTokenType() == PlSqlParserSUM ||
        lt1.GetTokenType() == PlSqlParserCOUNT ||
        lt1.GetTokenType() == PlSqlParserAVG ||
        lt1.GetTokenType() == PlSqlParserMIN ||
        lt1.GetTokenType() == PlSqlParserMAX ||
        lt1.GetTokenType() == PlSqlParserROUND ||
        lt1.GetTokenType() == PlSqlParserLEAST ||
        lt1.GetTokenType() == PlSqlParserGREATEST) &&
        lt2.GetTokenType() == PlSqlParserLEFT_PAREN {
        return false
    }
    return true
}

func (p *PlSqlParserBase) isNotStartOfJoin() bool {
    stream := p.GetTokenStream().(*antlr.CommonTokenStream)
    lt1 := stream.LT(1)
    if lt1.GetTokenType() == PlSqlParserINNER ||
       lt1.GetTokenType() == PlSqlParserCROSS ||
       lt1.GetTokenType() == PlSqlParserNATURAL ||
       lt1.GetTokenType() == PlSqlParserPARTITION ||
       lt1.GetTokenType() == PlSqlParserFULL ||
       lt1.GetTokenType() == PlSqlParserLEFT ||
       lt1.GetTokenType() == PlSqlParserRIGHT ||
       lt1.GetTokenType() == PlSqlParserOUTER {
        return false
    }
    return true
}

File: sql/plsql/Dart/PlSqlParserBase.dart
import 'package:antlr4/antlr4.dart';
import 'dart:io';
import 'dart:convert';
import 'PlSqlLexer.dart';

abstract class PlSqlParserBase extends Parser
{
    bool _isVersion12 = true;
    bool _isVersion11 = true;
    bool _isVersion10 = true;

    bool _lastUnitWasPlsql = false;

    PlSqlParserBase(TokenStream input) : super(input)
    {
    }

    @override
    void reset([bool resetInput = true]) {
        _lastUnitWasPlsql = false;
        super.reset(resetInput);
    }

    void setLastUnitPlsql() { _lastUnitWasPlsql = true; }
    void setLastUnitSql()   { _lastUnitWasPlsql = false; }
    bool isLastUnitSql()    { return !_lastUnitWasPlsql; }
    bool isLastUnitPlsql()  { return _lastUnitWasPlsql; }

    bool isSolidusSeparator() {
        var stream = tokenStream as CommonTokenStream;
        var solidus = stream.LT(1);
        if (solidus == null || solidus!.type != PlSqlLexer.TOKEN_SOLIDUS)
            return false;

        int solidusLine = solidus!.line;

        var prev = stream.LT(-1);
        if (prev != null && prev!.type != Token.EOF && prev.line == solidusLine)
            return false;

        var next = stream.LT(2);
        if (next != null && next!.type != Token.EOF && next.line == solidusLine)
            return false;

        return true;
    }

    bool isVersion12() {
        return _isVersion12;
    }

    void setVersion12(bool value) {
        _isVersion12 = value;
    }

    bool isVersion11() {
        return _isVersion11;
    }

    void setVersion11(bool value) {
        _isVersion11 = value;
    }

    bool isVersion10() {
        return _isVersion10;
    }

    void setVersion10(bool value) {
        _isVersion10 = value;
    }

    bool IsNotNumericFunction() {
        var stream = tokenStream as CommonTokenStream;
        var lt1 = stream.LT(1);
        var lt2 = stream.LT(2);
        if ((lt1!.type == PlSqlLexer.TOKEN_SUM ||
             lt1.type == PlSqlLexer.TOKEN_COUNT ||
             lt1.type == PlSqlLexer.TOKEN_AVG ||
             lt1.type == PlSqlLexer.TOKEN_MIN ||
             lt1.type == PlSqlLexer.TOKEN_MAX ||
             lt1.type == PlSqlLexer.TOKEN_ROUND ||
             lt1.type == PlSqlLexer.TOKEN_LEAST ||
             lt1.type == PlSqlLexer.TOKEN_GREATEST) && lt2!.type == PlSqlLexer.TOKEN_LEFT_PAREN)
            return false;
        return true;
    }

    bool isNotStartOfJoin() {
        var stream = tokenStream as CommonTokenStream;
        var lt1 = stream.LT(1);
        if (lt1!.type == PlSqlLexer.TOKEN_INNER ||
            lt1.type == PlSqlLexer.TOKEN_CROSS ||
            lt1.type == PlSqlLexer.TOKEN_NATURAL ||
            lt1.type == PlSqlLexer.TOKEN_PARTITION ||
            lt1.type == PlSqlLexer.TOKEN_FULL ||
            lt1.type == PlSqlLexer.TOKEN_LEFT ||
            lt1.type == PlSqlLexer.TOKEN_RIGHT ||
            lt1.type == PlSqlLexer.TOKEN_OUTER)
            return false;
        return true;
    }
}
